### ISSUE TITLE
nullable and oneOf/anyOf/etc

### ISSUE THREAD

**Issue by jayvdb** [pglass/py-openapi-schema-to-json-schema#10]
The converter ignores nullable here:

```py
    schema = {
        "oneOf": [
            {
                "type": "object",
                "required": ["foo"],
                "properties": {
                    "foo": {
                        "type": "integer",
                    },
                },
            },
            {
                "type": "object",
                "required": ["bar"],
                "properties": {
                    "bar": {
                        "type": "string",
                    },
                },
            },
        ],
        "nullable": True,
    }
```

https://spec.openapis.org/oas/v3.0.3#fixed-fields-19

> (nullable) A true value adds "null" to the allowed type specified by the type keyword, only if type is explicitly defined within the same Schema Object.

So the `null` cant be added to the top level `type` that doesnt exist.

I guess this requires special handling that groks oneOf, anyOf, etc and adds another schema to those lists with `type: null`.

**Comment by jayvdb** [comment]
I sort of expect that `nullable: true` is actually supposed to be invalid if used with `oneOf`, `anyOf`, etc., and not used with `type`.

The OpenAPI 3 spec doesnt clearly state this, however it suggests that `nullable` is ignored if `type` is not present.

However I dont see any explicit mention of that in the OpenAPI spec, or anything in the JSON Schema spec which might dictate that.  Thoughts?

If it is invalid, should this converter emit a warning, or an exception?

**Comment by pglass** [comment]
> A true value adds "null" to the allowed type specified by the type keyword, only if type is explicitly defined within the same Schema Object.

It is pretty clear that `nullable` only applies when `type` is present at the same level, so it's okay that it's ignored when `type` is not present. The source library, https://github.com/openapi-contrib/openapi-schema-to-json-schema, does the same.

A workaround is to add `nullable` to any of the schemas `oneOf` list. This doesn't work well when using `oneOf` with a list of `$ref`s though.

-----

The current problem is, I don't think there's a way to write the OpenAPI bit so that the converted JSON Schema accepts `null` when using `oneOf`.

For example, OpenAPI allows a `type: object` alongside the `oneOf` (I think):

```
{
    "type": "object",
    "oneOf": [
        ...
    ],
    "nullable": true
}
```

This is converted to the following JSON Schema, with a `type` array...

```
{
  "type": ["object", "null"],
  "oneOf": [
    ...
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}
```

But the above doesn't seem right, because actual JSON Schema validator implementations do not seem to accept null for that schema.

Instead, we need some `oneOf`-specific behavior. We could output the following, with `{"type": "null"}` added as an alternative in the `oneOf` list:

```
{
  "type": ["object", "null"], 
  "oneOf": [
    ...,
    {"type": "null"}
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}
```

Alternatively, we could support `nullable` without `type`, and add `{"type": "null"}` in the `oneOf` list.

```
{
  "oneOf": [
    ...,
    {"type": "null"}
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}
```

**Comment by pglass** [comment]
Same issue on the original library I ported from: https://github.com/mikunn/openapi-schema-to-json-schema/issues/31

There's a PHP library https://github.com/BenMorel/openapi-schema-to-json-schema which fixed the issue in https://github.com/BenMorel/openapi-schema-to-json-schema/commit/bb18eec80dcace69872010635f10ef511650e408

It looks like they take the latter approach I mentioned, which is to support `nullable` without `type` in a `oneOf`. They add `{"type": "null"}` into the `oneOf` list.

I'd accept that. This is the more usable option, though perhaps does not match the spec exactly (but perhaps so).

**Comment by jayvdb** [comment]
I also encountered that validators fail to handle `oneOf: [{"type": "null"}, ...]`

I had an idea last night that I have not yet tried with validators: inject a level

```yaml
type: ["object", "null"], 
oneOf: # injected level
  oneOf: # orig schema
    - {} # real alternative
    - {} # other real alternative
```

Another approach which I suspect will hit 'not/badly implemented' problems in validators: `if` , `then`

https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.2.2.1

```yaml
if: {..}  # constructed schema to match any type other than "null"
then: # injected level
  oneOf:  # orig schema
    - {} # real alternative
    - {} # other real alternative
```

**Comment by pglass** [comment]
> I also encountered that validators fail to handle oneOf: [{"type": "null"}, ...]

Hmm, do you have an example? I don't see the same behavior with any of the Python JSON Schema validators ([jsonschema](https://github.com/Julian/jsonschema), [fastjsonschema](https://horejsek.github.io/python-fastjsonschema/), and [jsonschema_rs](https://github.com/Stranger6667/jsonschema-rs/tree/master/python)) (with the simple schemas I tested, at least).

Those validators are consistent in accepting `null` for all of these schemas:

```json
{
  "oneOf": [{ "type": "null" }],
  "$schema": "http://json-schema.org/draft-04/schema#"
}

{
  "oneOf": [
    { "type": "null" },
    { "type": "object" }
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}

{
  "type": [ "object", "null" ],
  "oneOf": [
    { "type": "null" },
    { "type": "object" }
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}
```

The "oddball" is when `type: object` is at the top-level with `oneOf: [{type: null}, ...]`. All three validators reject `null` with this schema, (I guess because the top-level `type: object` must be satisfied in addition to the `oneOf`).

```json
{
  "type": "object",
  "oneOf": [
    { "type": "null" },
    { "type": "object" }
  ],
  "$schema": "http://json-schema.org/draft-04/schema#"
}

**Comment by pglass** [comment]
The anyOf/oneOf situation with nullable is fixed in #12, and released in [0.0.3](https://pypi.org/project/py-openapi-schema-to-json-schema/0.0.3/).

I opened #15 for the allOf+nullable case.
